imap/http_h2.c

```diff
@@ -749,15 +746,13 @@ static int resp_body_chunk(struct transaction_t *txn,
 
     /* NOTE: The protstream that we use as the data source MUST remain
        available until the data source read callback has retrieved all data.
-       Also note that we need to make a copy of the data because data frames
-       may not be sent prior to the original pointer becoming invalid.
     */
-    struct protstream *s = prot_readmap(xmemdup(data, datalen), datalen);
-    s->buf = s->ptr;  /* So prot_free() will free the copy of the data */
-
-    prd.source.ptr = s;
-    prd.read_callback = data_source_read_cb;
+    nghttp2_data_provider prd = {
+        .source.ptr = prot_readmap(data, datalen),
+        data_source_read_cb
+    };
```

Comment:
Style nit only, but I think I like it better when, if any struct fields are initialised by name, then the rest that are being initialised are also specified by name.

```suggestion
    nghttp2_data_provider prd = {
        .source.ptr = prot_readmap(data, datalen),
        .read_callback = data_source_read_cb,
    };
```

Though I think the way you wrote it is actually a positional initialisation of all fields, and .source.ptr is only named because source is a union rather than a struct, and you need to specify which member you're initialising.  Another way you might initialise the outer struct positionally is using a C99 compound literal that names the union member for the source field, something like this (untested...):

```suggestion
    nghttp2_data_provider prd = {
        (nghttp2_data_source){ .ptr = prot_readmap(data, datalen) },
        data_source_read_cb,
    };
```

But we don't presently make much (any?) use of C99 compound literals in Cyrus, and I've noticed in my own projects that editor syntax highlighting can find them confusing, so I present this option mostly as a point of interest.

Reply:
Added your first suggestion.